### PR TITLE
Skip empty text_vec examples in seq2seq.

### LINKED ISSUE #1592: Empty training examples throwing runtime error
I'm getting an error with the OpenSubtitles (2009 and 2018) and the Cornell Movie datasets with length 0 examples being loaded. My guess is that there are empty messages in the datasets, which could easily be removed with some preprocessing or a flag to avoid such examples. However, I'm having trouble finding such a flag and would appreciate any help you can provide. I do not have this issue arise on other datasets that I have explored so far. 

I'm a bit puzzled as to why others haven't seemed to run into this problem and was hoping you could help point me in the right direction, especially if I'm missing an easy fix. Thanks.

**Error:**
```
Traceback (most recent call last):
 File "examples/train_model.py", line 16, in <module>
   TrainLoop(opt).train()
 File "/data/odemasi/packages/ParlAI/parlai/scripts/train_model.py", line 523, in train
   world.parley()
 File "/data/odemasi/packages/ParlAI/parlai/core/worlds.py", line 654, in parley
   batch_act = self.batch_act(agent_idx, batch_observations[agent_idx])
 File "/data/odemasi/packages/ParlAI/parlai/core/worlds.py", line 627, in batch_act
   batch_actions = a.batch_act(batch_observation)
 File "/data/odemasi/packages/ParlAI/parlai/core/torch_agent.py", line 1206, in batch_act
   output = self.train_step(batch)
 File "/data/odemasi/packages/ParlAI/parlai/core/torch_generator_agent.py", line 478, in train_step
   raise e
 File "/data/odemasi/packages/ParlAI/parlai/core/torch_generator_agent.py", line 454, in train_step
   scores, preds, _ = self.model(batch.text_vec, batch.label_vec)
 File "/data/odemasi/anaconda3/envs/parlai_venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 489, in __call__
   result = self.forward(*input, **kwargs)
 File "/data/odemasi/packages/ParlAI/parlai/core/torch_generator_agent.py", line 214, in forward
   encoder_states = prev_enc if prev_enc is not None else self.encoder(xs)
 File "/data/odemasi/anaconda3/envs/parlai_venv/lib/python3.7/site-packages/torch/nn/modules/module.py", line 489, in __call__
   result = self.forward(*input, **kwargs)
 File "/data/odemasi/packages/ParlAI/parlai/agents/seq2seq/modules.py", line 226, in forward
   xes = pack_padded_sequence(xes, x_lens, batch_first=True)
 File "/data/odemasi/anaconda3/envs/parlai_venv/lib/python3.7/site-packages/torch/nn/utils/rnn.py", line 148, in pack_padded_sequence
   return PackedSequence(torch._C._VariableFunctions._pack_padded_sequence(input, lengths, batch_first))
RuntimeError: Length of all samples has to be greater than 0, but found an element in 'lengths' that is <= 0
```

**Resulted from:**
python examples/train_model.py \
-t cornell_movie \
-bs 32 \
--hiddensize 1024 \
--dict-minfreq 2 \
--dict-lower True \
--dict-file cornell_dict \
-m seq2seq \
-mf cornell_seq2seq

### PR DESCRIPTION
**Pull request by stephenroller**
Fixes #1592.

Can't just add the `__START__` token everywhere for backwards compatibility. Datasets should be adding a `__SILENCE__` token equivalent, but they don't always do.

This might be affecting transformer/generator too. Haven't checked.

TODO:
- [x] check transformer/generator
- [x] add an integration test

### CODE REVIEW

**Comment by stephenroller** [comment]
Data test fails intentionally. Point is to explicitly create a dataset with bad examples :)
```
devfair0237 parlai skipempties » python parlai/scripts/verify_data.py -t integration_tests:bad_example
[creating task(s): integration_tests:bad_example]
parlai/scripts/verify_data.py:52: UserWarning: warning: empty string label_candidate:

  warn_once(txt)
parlai/scripts/verify_data.py:52: UserWarning: warning: label missing in candidate_labels:

  warn_once(txt)
parlai/scripts/verify_data.py:52: UserWarning: warning: missing text field:

  warn_once(txt)
parlai/scripts/verify_data.py:52: UserWarning: warning: missing labels/eval_labels field:

  warn_once(txt)
[ loaded 500 episodes with a total of 500 examples ]
0s elapsed: {'exs': 438, '%done': '87.60%', 'time_left': '0s', 'missing_text': 62, 'missing_labels': 62, 'missing_label_candidates': 63, 'empty_label_candidates': 63, 'label_candidates_with_missing_label': 187}
```

File: parlai/tasks/integration_tests/agents.py [hunk 1]
```diff
@@ -171,5 +172,59 @@ def setup_data(self, fold):
...
+        4. label candidates is empty
+        5. label candidates contains an empty string
+        6. label isn't in the candidates
+        7. missing label candidates:
+    """
+    NUM_CASES = 8
+
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+        # gross hack: override sub thing to force things the way we want
```
Comment (emilydinan) [inline parlai/tasks/integration_tests/agents.py]: what is "sub thing"?

File: parlai/tasks/integration_tests/agents.py [hunk 2]
```diff
@@ -171,5 +172,59 @@ def setup_data(self, fold):
...
+    violated:
+
+        0. text is empty string
+        1. missing text
+        2. label is empty string
+        3. missing label
+        4. label candidates is empty
+        5. label candidates contains an empty string
+        6. label isn't in the candidates
+        7. missing label candidates:
```
Comment (emilydinan) [inline parlai/tasks/integration_tests/agents.py]: nit: what is this colon for?

File: parlai/tasks/integration_tests/agents.py [hunk 3]
```diff
@@ -171,5 +172,59 @@ def setup_data(self, fold):
...
+        # gross hack: override sub thing to force things the way we want
+        self.data.get = self._wrapperfn(self.data.get)
+
+    def _wrapperfn(self, oldget):
+        def newget(*args):
+            item, eod = oldget(*args)
+            item = copy.deepcopy(item)
+            newget.case = (newget.case + 1) % self.NUM_CASES
+            case = newget.case
+            # print(self.case)
```
Comment (emilydinan) [inline parlai/tasks/integration_tests/agents.py]: left this in here

File: parlai/tasks/integration_tests/agents.py [hunk 4]
```diff
@@ -171,5 +172,59 @@ def setup_data(self, fold):
...
+            elif case == 3:
+                del item['labels']
+            elif case == 4:
+                item['label_candidates'] = []
+            elif case == 5:
+                item['label_candidates'] = tuple(list(item['label_candidates']) + [''])
+            elif case == 6:
+                item['label_candidates'] = list(item['label_candidates'])
+                item['label_candidates'].remove(item['labels'][0])
+                item['label_candidates'] = tuple(item['label_candidates'])
```
Comment (emilydinan) [inline parlai/tasks/integration_tests/agents.py]: is there a reason we need to transform this back into a tuple?

**Comment by stephenroller** [comment]
Yeah I agree, data sets shouldn't produce null input, but some do. I personally don't mind us marking this as "undefined behavior", but still suspect things shouldn't *crash*, or if they do, they should with reasonable errors.

Anyway, the fact that datasets *sometimes* have these behaviors was enough for me to want to have a test that things don't break like they do in #1592.

**Pull request merged**